cou/apps/auxiliary.py

```diff
@@ -502,3 +508,156 @@ async def _verify_nova_compute(self, target: OpenStackRelease) -> None:
+
+        :return: Temporary file path
+        :rtype: str
+        """
+        cacert_b64 = self.config["ssl-ca"].get("value")
+        cacert_file = None
+        if cacert_b64:
+            with tempfile.NamedTemporaryFile(mode="wb", delete=False) as fp:
+                fp.write(base64.b64decode(cacert_b64))
+                cacert_file = fp.name
+                logger.debug("Create tempfile: %s", cacert_file)
+        return cacert_file
+
+    async def _get_unit_api_url(self, unit_name: str) -> str:
+        """Get unit's vault api address.
+
+        :param unit_name: vault unit name
+        :type unit_name: str
+        :return: unit's vault api url
+        :rtype: str
+        """
+        if self.config["hostname"].get("value"):  # Use hostname if hostname is being used.
+            address = self.config["hostname"].get("value")
+        elif self.config["vip"].get("value"):  # Use vip as address if vip is being used.
+            address = self.config["vip"].get("value")
+        else:
+            juju_unit = await self.model.get_unit(unit_name)
+            address = juju_unit.public_address
+
+        transport = "https" if self.config["ssl-cert"].get("value") else "http"
+        return f"{transport}://{address}:8200"
+
+    async def _wait_for_sealed_status(self) -> None:
+        """Wait for application vault go into sealed status.
+
+        :raises ApplicationError: When application vault is not in sealed status.
+        """
+        await self.model.wait_for_idle(
+            timeout=self.wait_timeout,
+            status="blocked",
+            apps=[self.name],
+            # Vault application will go into error status first then blocked status.
+            raise_on_error=False,
+        )
+
+        app_status = await self.model.get_application_status(app_name=self.name)
+        if not app_status.status.info == "Unit is sealed":
+            raise ApplicationError("Vault not in sealed status")
+        logger.debug("Application 'vault' in sealed status")
+
+    async def _get_vault_client(self, unit_name: str, cacert_file: Optional[str]) -> hvac.Client:
+        """Get vault client.
+
+        :param unit_name: vault unit name
+        :type unit_name: str
+        :param cacert_file: cacert file path
+        :type cacert_file: str
+        :return: hvac vault Client object
+        :rtype: hvac.Client
+        """
+        vault_url = await self._get_unit_api_url(unit_name)
+        client = hvac.Client(url=vault_url, verify=cacert_file)
+        return client
+
+    async def _unseal_vault(self) -> None:
+        """Unseal vault on every vault unit."""
+        # Stop progress_indicator because it will clear the unseal key input.
+        progress_indicator.stop()
+
+        for unit_name in self.units:
+            logger.debug("Start unseal %s", unit_name)
+            cacert_file = self._get_cacert_file()
+            client = await self._get_vault_client(unit_name, cacert_file)
+            while True:
```

Comment:
I think the UX could be improved here - could we add some more feedback to the user in the loop?  For example:

- a message first to give instructions to the user (vault is sealed, you need to enter the unseal keys, cou will ask for unseal keys one at a time until vault is unsealed, blank lines are ignored)
- if a blank line is entered (with the current logic), a message should be printed that it was ignored (so the user doesn't worry)
- when vault is still sealed after entering an unseal key, tell the user that the unseal key was submitted successfully, but vault is still sealed (try entering more keys)
- when vault is successfully unsealed, print a success message
- also can we print the number of keys required? and the number of keys remaining to be entered?

Reply:
Can this one be a separate one? I would like the limit the scope here.